curp/src/server.rs

```diff
@@ -764,6 +843,44 @@ impl<C: 'static + Command, CE: 'static + CommandExecutor<C>> Protocol<C, CE> {
             state.term,
         )))
     }
+
+    /// Handle `Vote` requests
+    #[allow(clippy::pedantic)] // need not return result, but to keep it consistent with rpc handler functions, we keep it this way
+    fn vote(
+        &self,
+        request: tonic::Request<VoteRequest>,
+    ) -> Result<tonic::Response<VoteResponse>, tonic::Status> {
+        let req = request.into_inner();
+        debug!(
+            "vote received: term({}), last_log_index({}), last_log_term({}), id({})",
+            req.term, req.last_log_index, req.last_log_term, req.candidate_id
+        );
+
+        let state = self.state.upgradable_read();
+        if req.term < state.term || state.voted_for.is_some() {
```

Comment:
Even if you voted_for someone, you should also accept a vote with a higher term.